Alex and Happy Part 2/article.md

```diff
@@ -0,0 +1,1372 @@
+* Conditional expressions: if-then-else and if-then.
+
+In addition, exercises will guide you through extending the language with patterns, list access, and pattern matching. The first part also introduced block comments (in the main text) and rational numbers and line comments (in exercises).
+
+## Happy vs. Megaparsec
+
+Happy is a parser generator that generates LALR(1) (bottom-up) parsers, which is a more performant yet less powerful version of LR(1) parsers.
+
+Meanwhile, Haskell libraries such as Megaparsec work best on LL(1) or LL(∞) grammars. They're implemented as recursive descent parsers (top-down parsers) with backtracking.
+
+As such, these libraries have caveats like not being able to handle left-recursion. Their performance can also exponentially degrade due to backtracking.
+
+While comparing the implementations of both parsing techniques is beyond the scope of this post, you can read two suggested blog posts for more information: [From recursive descent to LR parsing](https://www.abubalay.com/blog/2021/12/31/lr-control-flow) and [Which Parsing Approach?](https://tratt.net/laurie/essays/entries/which_parsing_approach.html).
+
+Another difference between these libraries is that Happy is a parser generator.
+It means that instead of writing Haskell files (`.hs`) like you do with Megaparsec, you will write Happy (`.y`) files. Those will generate Haskell files with implementations of the parser recognizing the described grammar.
+
+To sum it up, here are some pros and cons of Happy in comparison to Megaparsec:
+
+Pros:
+* Parser generators report ambiguous grammars to the user, creating greater trust in the parser's output and saving from debugging headaches.
+* Bottom-up parsers can handle left-recursion, while top-down parsers can't.
+* The performance will be more predictable, as Megaparsec can cause exponential time complexity with the use of `try`. In addition, Happy's "static" generated LALR algorithm has better known complexity constraints.
+* Dealing with operator precedence and associativity requires less labor. In Megaparsec, there are standard tools like [`makeExprParser`](https://hackage.haskell.org/package/parser-combinators-1.3.0/docs/Control-Monad-Combinators-Expr.html#v:makeExprParser) to deal with them, but once you need something that is non-standard, you'll need to write more non-trivial parser rules by hand.
+
+Cons:
+* Less idiomatic, as you need to write Happy grammars instead of Haskell code.
+* Less flexible since you can't change Happy’s algorithm. Meanwhile, parser combinators enable you to easily create your own functions.
+
+## How it works
+
+An in-depth explanation of how an LR(1) parser works is well beyond the scope of this tutorial. For this, we recommend you read our article on [how to Implement an LR(1) Parser](https://serokell.io/blog/how-to-implement-lr1-parser).
+We will, however, introduce a few basic concepts required to understand this article better.
+
+### Terminology
```

Comment:
With a little bit of effort, this can be a story not a terminology.

Reply:
Could you elaborate on this?

Reply:
E.g. imagine if the heading was:

### High-level overview of parsing 

and what would be the contents.

Reply:
OK, I rewrote this with a more prose-like style, and I also used a concrete example that should hopefully make things clearer.

Reply:
Hmm, I just saw your story, it's pretty nice, I'll rewrite this part again so I can use it. Or would it be better to not have Happy examples between the story?

Reply:
As a general rule, it's good to have examples, period 🙂 I honestly would also include an example with on-paper derivation and draw parallels to tree building, but tbh what you have currently is good enough (aside from a couple small nitpicks), the article isn't about "a complete overly pedantic introduction to formal language theory" after all😅